test/e2e/storage/dax/dax.go

```diff
@@ -0,0 +1,247 @@
+				},
+			},
+			RestartPolicy: v1.RestartPolicyNever,
+			NodeSelector:  config.ClientNodeSelector,
+			NodeName:      config.ClientNodeName,
+		},
+	}
+	switch volumeMode {
+	case v1.PersistentVolumeBlock:
+		// This is what we would like to use:
+		//
+		// pod.Spec.Containers[0].VolumeDevices = append(pod.Spec.Containers[0].VolumeDevices,
+		// 	v1.VolumeDevice{
+		// 		Name:       volName,
+		// 		DevicePath: "/dax-dev",
+		// 	})
+
+		// But because of https://github.com/kubernetes/kubernetes/issues/85624, /dax-dev
+		// then is silently ignored.
+		//
+		// Instead we have to use the workaround mentioned in that issue:
+		// - bring up an unprivileged init container with /dax-dev and a shared empty volume on /data
+		// - get major/minor number of that device and put it into a script
+		// - re-create the device in the privileged container with that script
+		emptyDirName := "data"
+		pod.Spec.InitContainers = []v1.Container{
+			{
+				Name:  "copy-dax-dev",
+				Image: os.Getenv("PMEM_CSI_IMAGE"),
+				Command: []string{"sh", "-c",
+					"(echo '#!/bin/sh' && stat --format 'mknod /dax-dev b 0x%t 0x%T' /dax-dev) >/data/create-dax-dev.sh && chmod a+x /data/create-dax-dev.sh",
+				},
+				VolumeMounts: []v1.VolumeMount{
+					{
+						Name:      emptyDirName,
+						MountPath: "/data",
+					},
+				},
+				VolumeDevices: []v1.VolumeDevice{
+					{
+						Name:       volName,
+						DevicePath: "/dax-dev",
+					},
+				},
+			},
+		}
+		pod.Spec.Containers[0].VolumeMounts = append(pod.Spec.Containers[0].VolumeMounts,
+			v1.VolumeMount{
+				Name:      emptyDirName,
+				MountPath: "/data",
+			},
+		)
+		pod.Spec.Volumes = append(pod.Spec.Volumes, v1.Volume{
+			Name: emptyDirName,
+			VolumeSource: v1.VolumeSource{
+				EmptyDir: &v1.EmptyDirVolumeSource{},
+			},
+		})
+	default:
+		pod.Spec.Containers[0].VolumeMounts = append(pod.Spec.Containers[0].VolumeMounts,
+			v1.VolumeMount{
+				Name:      volName,
+				MountPath: "/mnt",
+			},
+		)
+	}
+
+	By(fmt.Sprintf("Creating pod %s", pod.Name))
+	ns := f.Namespace.Name
+	podClient := f.PodClientNS(ns)
+	createdPod := podClient.Create(pod)
+	defer func() {
+		By("delete the pod")
+		podClient.DeleteSync(createdPod.Name, &metav1.DeleteOptions{}, framework.DefaultPodDeletionTimeout)
+	}()
+	podErr := e2epod.WaitForPodRunningInNamespace(f.ClientSet, createdPod)
+	framework.ExpectNoError(podErr, "running pod")
+	if volumeMode == v1.PersistentVolumeBlock {
+		By("mounting raw block device")
+		// TODO: remove the workaround above and script invocation here.
+		pmempod.RunInPod(f, l.root, nil, "/data/create-dax-dev.sh && mkfs.ext4 -b 4096 /dax-dev && mkdir -p /mnt && mount -odax /dax-dev /mnt", ns, pod.Name, containerName)
+	}
+
+	By("checking that missing DAX support is detected")
+	pmempod.RunInPod(f, l.root, []string{l.daxCheckBinary}, l.daxCheckBinary+" /no-dax; if [ $? -ne 1 ]; then echo should have reported missing DAX >&2; exit 1; fi", ns, pod.Name, containerName)
```

Comment:
@pohly I couldn't get this point. Where we are copying the test binary to the target container? Can you please help me to understand.

Reply:
It's done by `RunInPod`, see https://github.com/intel/pmem-csi/blob/46259b59ace37e69b2678a1500a65c2372e2d069/test/e2e/pod/exec.go#L24-L41

Reply:
Oops, sorry I was bit lazy :). Thanks.